app/Http/Controllers/Api/ExperienceSkillsController.php

```diff
@@ -48,4 +49,70 @@ public function destroy(ExperienceSkill $experienceSkill)
     {
         $experienceSkill->delete();
     }
+
+    public function batchStore(BatchStoreExperienceSkill $request)
+    {
+        $newExperienceSkills = $request->validated();
+        $response = [];
+
+        foreach ($newExperienceSkills as $newExperienceSkill) {
+            // Restore soft deleted experienceSkill if it exists, otherwise create a new one.
+            $softDeletedExperienceSkill = ExperienceSkill::withTrashed()
+                ->where([
+                ['skill_id', $newExperienceSkill['skill_id']],
+                ['experience_id', $newExperienceSkill['experience_id']],
+                ['experience_type', $newExperienceSkill['experience_type']]
+            ])->first();
+            if ($softDeletedExperienceSkill) {
+                $softDeletedExperienceSkill->restore();
+                if ($newExperienceSkill['justification'] !== null && $newExperienceSkill['justification'] !== '') {
+                    $softDeletedExperienceSkill->justification = $newExperienceSkill['justification'];
+                }
+                $softDeletedExperienceSkill->save();
+                array_push($response, $softDeletedExperienceSkill);
+            } else {
+                $experienceSkill = new ExperienceSkill($newExperienceSkill);
+                $experienceSkill->skill_id = $newExperienceSkill['skill_id'];
+                $experienceSkill->experience_id = $newExperienceSkill['experience_id'];
+                $experienceSkill->experience_type = $newExperienceSkill['experience_type'];
+                $experienceSkill->justification = $newExperienceSkill['justification'];
+                $experienceSkill->save();
+                $experienceSkill->fresh();
+                array_push($response, $experienceSkill);
+            }
+        }
+
+        return new JsonResource($response);
+    }
+
+    public function batchUpdate(Request $request)
+    {
+        $newExperienceSkills = $request->validate([
+            '*.id' => 'required|exists:App\Models\ExperienceSkill,id',
+            '*.justification' => 'nullable|string',
+        ]);
+        $response = [];
+
+        foreach ($newExperienceSkills as $newExperienceSkill) {
+            $experienceSkill = ExperienceSkill::where('id', $newExperienceSkill['id'])->first();
```

Comment:
@Khristophor also had some optimization suggestions for this. You can get all the ExperienceSkills with one query outside of the loop with `ExperienceSkills::whereIn('id', $newExperienceSkills ->pluck('id')->all())->get()`. We want to minimize direct DB manipulation within the loop.